File: tests/wasm/test_roc.js
/**
 * WASM binding tests for ROC indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('ROC partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.roc_js(close, 9);
    assert.strictEqual(result.length, close.length);
});

test('ROC accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.roc;
    
    const result = wasm.roc_js(
        close,
        expected.defaultParams.period
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-7,
        "ROC last 5 values mismatch"
    );
    
    
    await compareWithRust('roc', result, 'close', expected.defaultParams);
});

test('ROC default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.roc_js(close, 9);
    assert.strictEqual(result.length, close.length);
});

test('ROC zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.roc_js(inputData, 0);
    }, /Invalid period/);
});

test('ROC period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.roc_js(dataSmall, 10);
    }, /Invalid period/);
});

test('ROC very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.roc_js(singlePoint, 9);
    }, /Invalid period|Not enough valid data/);
});

test('ROC empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.roc_js(empty, 9);
    }, /Empty data provided|Input data slice is empty/);
});

test('ROC reinput', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const firstResult = wasm.roc_js(close, 14);
    assert.strictEqual(firstResult.length, close.length);
    
    
    const secondResult = wasm.roc_js(firstResult, 14);
    assert.strictEqual(secondResult.length, firstResult.length);
    
    
    for (let i = 28; i < secondResult.length; i++) {
        assert(!isNaN(secondResult[i]), 
               `Expected no NaN after index 28, found NaN at ${i}`);
    }
});

test('ROC NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.roc_js(close, 9);
    assert.strictEqual(result.length, close.length);
    
    
    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
    
    
    assertAllNaN(result.slice(0, 9), "Expected NaN in warmup period");
});

test('ROC all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.roc_js(allNaN, 10);
    }, /All values are NaN/);
});

test('ROC batch single parameter set', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const batchResult = wasm.roc_batch(close, {
        period_range: [10, 10, 0]
    });
    
    
    const singleResult = wasm.roc_js(close, 10);
    
    assert.strictEqual(batchResult.values.length, singleResult.length);
    assertArrayClose(batchResult.values, singleResult, 1e-10, "Batch vs single mismatch");
});

test('ROC batch multiple periods', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const batchResult = wasm.roc_batch(close, {
        period_range: [10, 14, 2]  
    });
    
    
    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);
    
    
    const periods = [10, 12, 14];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.values.slice(rowStart, rowEnd);
        
        const singleResult = wasm.roc_js(close, periods[i]);
        assertArrayClose(
            rowData, 
            singleResult, 
            1e-10, 
            `Period ${periods[i]} mismatch`
        );
    }
});

test('ROC batch metadata from result', () => {
    
    const close = new Float64Array(20); 
    close.fill(100);
    
    const result = wasm.roc_batch(close, {
        period_range: [10, 14, 2]  
    });
    
    
    assert.strictEqual(result.combos.length, 3);
    
    
    assert.strictEqual(result.combos[0].period, 10);
    
    
    assert.strictEqual(result.combos[2].period, 14);
});

test('ROC batch full parameter sweep', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    const batchResult = wasm.roc_batch(close, {
        period_range: [10, 14, 2]  
    });
    
    
    assert.strictEqual(batchResult.combos.length, 3);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 50);
    assert.strictEqual(batchResult.values.length, 3 * 50);
    
    
    for (let combo = 0; combo < batchResult.combos.length; combo++) {
        const period = batchResult.combos[combo].period;
        
        const rowStart = combo * 50;
        const rowData = batchResult.values.slice(rowStart, rowStart + 50);
        
        
        for (let i = 0; i < period; i++) {
            assert(isNaN(rowData[i]), `Expected NaN at warmup index ${i} for period ${period}`);
        }
        
        
        for (let i = period; i < 50; i++) {
            assert(!isNaN(rowData[i]), `Unexpected NaN at index ${i} for period ${period}`);
        }
    }
});

test('ROC batch edge cases', () => {
    
    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
    
    
    const singleBatch = wasm.roc_batch(close, {
        period_range: [5, 5, 1]
    });
    
    assert.strictEqual(singleBatch.values.length, 15);
    assert.strictEqual(singleBatch.combos.length, 1);
    
    
    const largeBatch = wasm.roc_batch(close, {
        period_range: [5, 7, 10] 
    });
    
    
    assert.strictEqual(largeBatch.values.length, 15);
    assert.strictEqual(largeBatch.combos.length, 1);
    
    
    assert.throws(() => {
        wasm.roc_batch(new Float64Array([]), {
            period_range: [10, 10, 0]
        });
    }, /All values are NaN/);
});


test('ROC zero-copy API', () => {
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const period = 5;
    
    
    const ptr = wasm.roc_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');
    
    
    const memView = new Float64Array(
        wasm.__wasm.memory.buffer,
        ptr,
        data.length
    );
    
    
    memView.set(data);
    
    
    try {
        wasm.roc_into(ptr, ptr, data.length, period);
        
        
        const regularResult = wasm.roc_js(data, period);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue; 
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {
        
        wasm.roc_free(ptr, data.length);
    }
});

test('ROC zero-copy with large dataset', () => {
    const size = 100000;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = Math.sin(i * 0.01) + Math.random() * 0.1 + 100; 
    }
    
    const ptr = wasm.roc_alloc(size);
    assert(ptr !== 0, 'Failed to allocate large buffer');
    
    try {
        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        memView.set(data);
        
        wasm.roc_into(ptr, ptr, size, 10);
        
        
        const memView2 = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        
        
        for (let i = 0; i < 10; i++) {
            assert(isNaN(memView2[i]), `Expected NaN at warmup index ${i}`);
        }
        
        
        for (let i = 10; i < Math.min(100, size); i++) {
            assert(!isNaN(memView2[i]), `Unexpected NaN at index ${i}`);
        }
    } finally {
        wasm.roc_free(ptr, size);
    }
});


test('ROC zero-copy error handling', () => {
    
    assert.throws(() => {
        wasm.roc_into(0, 0, 10, 10);
    }, /null pointer|Null pointer/i);
    
    
    const ptr = wasm.roc_alloc(10);
    try {
        
        assert.throws(() => {
            wasm.roc_into(ptr, ptr, 10, 0);
        }, /Invalid period/);
        
        
        assert.throws(() => {
            wasm.roc_into(ptr, ptr, 10, 20);
        }, /Invalid period/);
    } finally {
        wasm.roc_free(ptr, 10);
    }
});


test('ROC zero-copy memory management', () => {
    
    const sizes = [100, 1000, 10000, 100000];
    
    for (const size of sizes) {
        const ptr = wasm.roc_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.roc_free(ptr, size);
    }
});

test.after(() => {
    console.log('ROC WASM tests completed');
});

File: tests/python/test_roc.py
"""
Python binding tests for ROC indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestRoc:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_roc_partial_params(self, test_data):
        """Test ROC with partial parameters (None values) - mirrors check_roc_partial_params"""
        close = test_data['close']


        result = ta_indicators.roc(close, 10)
        assert len(result) == len(close)

    def test_roc_accuracy(self, test_data):
        """Test ROC matches expected values from Rust tests - mirrors check_roc_accuracy"""
        close = test_data['close']

        result = ta_indicators.roc(close, period=10)

        assert len(result) == len(close)


        expected_last_five = [
            -0.22551709049294377,
            -0.5561903481650754,
            -0.32752013235864963,
            -0.49454153980722504,
            -1.5045927020536976,
        ]
        assert_close(
            result[-5:],
            expected_last_five,
            rtol=1e-8,
            msg="ROC last 5 values mismatch"
        )


        for i in range(9):
            assert np.isnan(result[i]), f"Expected NaN at index {i} during warmup"

    def test_roc_default_candles(self, test_data):
        """Test ROC with default parameters - mirrors check_roc_default_candles"""
        close = test_data['close']


        result = ta_indicators.roc(close, 10)
        assert len(result) == len(close)

    def test_roc_zero_period(self):
        """Test ROC fails with zero period - mirrors check_roc_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.roc(input_data, period=0)

    def test_roc_period_exceeds_length(self):
        """Test ROC fails when period exceeds data length - mirrors check_roc_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.roc(data_small, period=10)

    def test_roc_very_small_dataset(self):
        """Test ROC fails with insufficient data - mirrors check_roc_very_small_dataset"""
        single_point = np.array([42.0])

        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.roc(single_point, period=9)

    def test_roc_reinput(self, test_data):
        """Test ROC applied twice (re-input) - mirrors check_roc_reinput"""
        close = test_data['close']


        first_result = ta_indicators.roc(close, period=14)
        assert len(first_result) == len(close)


        second_result = ta_indicators.roc(first_result, period=14)
        assert len(second_result) == len(first_result)


        for i in range(28, len(second_result)):
            assert not np.isnan(second_result[i]), f"Expected value after warmup, but found NaN at index {i}"

    def test_roc_nan_handling(self, test_data):
        """Test ROC handles NaN values correctly - mirrors check_roc_nan_handling"""
        close = test_data['close']

        result = ta_indicators.roc(close, period=9)
        assert len(result) == len(close)


        if len(result) > 240:
            for i in range(240, len(result)):
                assert not np.isnan(result[i]), f"Found unexpected NaN at index {i}"

    def test_roc_all_nan_input(self):
        """Test ROC with all NaN values"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.roc(all_nan, period=10)

    def test_roc_empty_input(self):
        """Test ROC fails with empty input"""
        empty = np.array([])

        with pytest.raises(ValueError):
            ta_indicators.roc(empty, period=10)

    def test_roc_batch_single_parameter_set(self, test_data):
        """Test batch with single parameter combination"""
        close = test_data['close']


        batch_result = ta_indicators.roc_batch(
            close,
            period_range=(10, 10, 0)
        )


        single_result = ta_indicators.roc(close, 10)

        assert batch_result['values'].shape == (1, len(close))
        assert_close(
            batch_result['values'][0],
            single_result,
            rtol=1e-10,
            msg="Batch vs single mismatch"
        )


        assert len(batch_result['periods']) == 1
        assert batch_result['periods'][0] == 10

    def test_roc_batch_multiple_parameters(self, test_data):
        """Test batch with multiple parameter values"""
        close = test_data['close'][:100]


        batch_result = ta_indicators.roc_batch(
            close,
            period_range=(10, 12, 2)
        )


        assert batch_result['values'].shape == (2, 100)


        expected_periods = [10, 12]

        assert list(batch_result['periods']) == expected_periods


        single_result = ta_indicators.roc(close, 10)
        assert_close(
            batch_result['values'][0],
            single_result,
            rtol=1e-10,
            msg="First batch row mismatch"
        )

    def test_roc_batch_edge_cases(self):
        """Test edge cases for batch processing"""
        close = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=np.float64)


        single_batch = ta_indicators.roc_batch(
            close,
            period_range=(3, 3, 1)
        )

        assert single_batch['values'].shape == (1, 10)


        large_batch = ta_indicators.roc_batch(
            close,
            period_range=(3, 5, 10)
        )


        assert large_batch['values'].shape == (1, 10)
        assert large_batch['periods'][0] == 3

    def test_roc_streaming(self, test_data):
        """Test ROC streaming functionality"""
        close = test_data['close']


        stream = ta_indicators.RocStream(period=10)


        stream_results = []
        for price in close:
            result = stream.update(price)
            stream_results.append(result if result is not None else np.nan)


        batch_result = ta_indicators.roc(close, period=10)


        assert_close(
            stream_results,
            batch_result,
            rtol=1e-9,
            msg="Streaming vs batch mismatch"
        )

    def test_roc_kernel_parameter(self, test_data):
        """Test ROC with different kernel parameters"""
        close = test_data['close'][:100]


        kernels = [None, 'scalar', 'avx2', 'avx512', 'auto']

        for kernel in kernels:
            try:
                if kernel:
                    result = ta_indicators.roc(close, 10, kernel=kernel)
                else:
                    result = ta_indicators.roc(close, 10)
                assert len(result) == len(close)
            except ValueError as e:

                msg = str(e)
                allowed = (
                    "Unsupported kernel" in msg
                    or "not available on this CPU" in msg
                    or "not compiled in this build" in msg
                )
                if not allowed:
                    raise


if __name__ == "__main__":
    pytest.main([__file__])
